Refactor PlaybackSessionInterfaceAVKit

danae404 opened the pull request:
#### 8f3c7ed80e7d7aa61480065dec88bea27bf7c997
<pre>
Refactor PlaybackSessionInterfaceAVKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=268418">https://bugs.webkit.org/show_bug.cgi?id=268418</a>
<a href="https://rdar.apple.com/121965868">rdar://121965868</a>

Reviewed by Jer Noble.

This patch refactors PlaybackSessionInterfaceAVKit. It makes a base class, PlaybackSessionInterfaceIOS,
and two classes that are derived from it: PlaybackSessionInterfaceAVKit and PlaybackSessionInterfaceLMK.
All AVKit-specific code is in PlaybackSessionInterfaceAVKit. PlaybackSessionInterfaceLMK will be used on
the vision platform in a later patch. As for now, PlaybackSessionInterfaceAVKit is the only class that is ever
instantiated.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/PlatformPlaybackSessionInterface.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.mm:
(WebCore::PlaybackSessionInterfaceAVKit::PlaybackSessionInterfaceAVKit):
(WebCore::PlaybackSessionInterfaceAVKit::playerController const):
(WebCore::PlaybackSessionInterfaceAVKit::invalidate):
(WebCore::PlaybackSessionInterfaceAVKit::playbackSessionModel const): Deleted.
(WebCore::PlaybackSessionInterfaceAVKit::modelDestroyed): Deleted.
(WebCore::PlaybackSessionInterfaceAVKit::logIdentifier const): Deleted.
(WebCore::PlaybackSessionInterfaceAVKit::loggerPtr const): Deleted.
(WebCore::PlaybackSessionInterfaceAVKit::logChannel const): Deleted.
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.h: Copied from Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.h.
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.mm: Added.
(WebCore::PlaybackSessionInterfaceIOS::PlaybackSessionInterfaceIOS):
(WebCore::PlaybackSessionInterfaceIOS::playbackSessionModel const):
(WebCore::PlaybackSessionInterfaceIOS::modelDestroyed):
(WebCore::PlaybackSessionInterfaceIOS::logIdentifier const):
(WebCore::PlaybackSessionInterfaceIOS::loggerPtr const):
(WebCore::PlaybackSessionInterfaceIOS::logChannel const):
* Source/WebCore/platform/ios/PlaybackSessionInterfaceLMK.h: Copied from Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.h.
* Source/WebCore/platform/ios/PlaybackSessionInterfaceLMK.mm: Added.
(WebCore::PlaybackSessionInterfaceLMK::PlaybackSessionInterfaceLMK):
(WebCore::PlaybackSessionInterfaceLMK::~PlaybackSessionInterfaceLMK):
(WebCore::PlaybackSessionInterfaceLMK::durationChanged):
(WebCore::PlaybackSessionInterfaceLMK::currentTimeChanged):
(WebCore::PlaybackSessionInterfaceLMK::bufferedTimeChanged):
(WebCore::PlaybackSessionInterfaceLMK::rateChanged):
(WebCore::PlaybackSessionInterfaceLMK::seekableRangesChanged):
(WebCore::PlaybackSessionInterfaceLMK::canPlayFastReverseChanged):
(WebCore::PlaybackSessionInterfaceLMK::audioMediaSelectionOptionsChanged):
(WebCore::PlaybackSessionInterfaceLMK::legibleMediaSelectionOptionsChanged):
(WebCore::PlaybackSessionInterfaceLMK::externalPlaybackChanged):
(WebCore::PlaybackSessionInterfaceLMK::wirelessVideoPlaybackDisabledChanged):
(WebCore::PlaybackSessionInterfaceLMK::mutedChanged):
(WebCore::PlaybackSessionInterfaceLMK::volumeChanged):
(WebCore::PlaybackSessionInterfaceLMK::invalidate):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
(VideoPresentationInterfaceIOS::create):
(VideoPresentationInterfaceIOS::VideoPresentationInterfaceIOS):
* Source/WebCore/platform/ios/WebAVPlayerController.h:
* Source/WebCore/platform/ios/WebAVPlayerController.mm:
(-[WebAVPlayerController playbackSessionInterface]):
(-[WebAVPlayerController setPlaybackSessionInterface:]):
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
(VideoFullscreenControllerContext::setUpFullscreen):
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionManagerProxy::createModelAndInterface):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(WKFullScreenViewControllerPlaybackSessionModelClient::setInterface):
* Source/WebKitLegacy/mac/WebView/WebVideoFullscreenController.mm:
(-[WebVideoFullscreenController init]):

Canonical link: <a href="https://commits.webkit.org/273994@main">https://commits.webkit.org/273994@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa19e9baf02fc13669d1aaa1145dd9cf0e4a3aab

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37416 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16298 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39679 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39942 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33342 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38727 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18904 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13449 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31767 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37981 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13735 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32840 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11952 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11936 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33500 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41205 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33826 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/33938 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37831 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12502 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10071 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35987 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13976 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8441 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12913 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13292 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -0,0 +1,153 @@
+}
+
+void PlaybackSessionInterfaceLMK::audioMediaSelectionOptionsChanged(const Vector<MediaSelectionOption>& options, uint64_t selectedIndex)
+{
```
for this method and similar ones, can the parameter names just be omitted instead of using `UNUSED_PARAM`?

aestes commented on Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2016-2019 Apple Inc. All rights reserved.
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
We should maintain the earliest copyright year here, so `2016-2024`.

aestes commented on Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.mm:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2014-2019 Apple Inc. All rights reserved.
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
```
Same comment about copyright.

aestes commented on Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.mm:
```diff
@@ -50,14 +47,15 @@
     [m_playerController setPlaybackSessionInterface:this];
     [m_playerController setDelegate:&model];
 
+    model.addClient(*this);
```
Shouldn't this be done in `PlaybackSessionInterfaceIOS:: PlaybackSessionInterfaceIOS()`?

aestes commented on Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.mm:
```diff
@@ -84,9 +82,19 @@
+
+    [m_playerController setDelegate:nullptr];
+    m_playbackSessionModel->removeClient(*this);
+    m_playbackSessionModel = nullptr;
```
I think the last two lines here should be moved to `PlaybackSessionInterfaceIOS::invalidate()` since both subclasses need to do this.

aestes commented on Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.mm:
```diff
@@ -0,0 +1,91 @@
+}
+const char* PlaybackSessionInterfaceIOS::logClassName() const
+{
+    return "PlaybackSessionInterfaceIOS";
```
`logClassName()` should be virtual so we can tell the difference between PlaybackSessionInterfaceAVKit and PlaybackSessionInterfaceLMK in logs.

aestes commented on Source/WebCore/platform/ios/PlaybackSessionInterfaceLMK.mm:
```diff
@@ -0,0 +1,153 @@
+    audioMediaSelectionOptionsChanged(model.audioMediaSelectionOptions(), model.audioMediaSelectedIndex());
+    legibleMediaSelectionOptionsChanged(model.legibleMediaSelectionOptions(), model.legibleMediaSelectedIndex());
+    externalPlaybackChanged(model.externalPlaybackEnabled(), model.externalPlaybackTargetType(), model.externalPlaybackLocalizedDeviceName());
+    wirelessVideoPlaybackDisabledChanged(model.wirelessVideoPlaybackDisabled());
```
This is all duplicated from `PlaybackSessionInterfaceAVKit`, so it should move to the `PlaybackSessionInterfaceIOS` constructor.

danae404 replied:
I think because these functions are virtual in `PlaybackSessionInterfaceIOS` and implemented in the derived classes, they cannot be called from `PlaybackSessionInterfaceIOS`'s constructor

danae404 committed:
Refactor PlaybackSessionInterfaceAVKit
https://bugs.webkit.org/show_bug.cgi?id=268418
rdar://121965868

Reviewed by Jer Noble.

This patch refactors PlaybackSessionInterfaceAVKit. It makes a base class, PlaybackSessionInterfaceIOS,
and two classes that are derived from it: PlaybackSessionInterfaceAVKit and PlaybackSessionInterfaceLMK.
All AVKit-specific code is in PlaybackSessionInterfaceAVKit. PlaybackSessionInterfaceLMK will be used on
the vision platform in a later patch. As for now, PlaybackSessionInterfaceAVKit is the only class that is ever
instantiated.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/PlatformPlaybackSessionInterface.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.h:
* Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.mm:
(WebCore::PlaybackSessionInterfaceAVKit::PlaybackSessionInterfaceAVKit):
(WebCore::PlaybackSessionInterfaceAVKit::playerController const):
(WebCore::PlaybackSessionInterfaceAVKit::invalidate):
(WebCore::PlaybackSessionInterfaceAVKit::playbackSessionModel const): Deleted.
(WebCore::PlaybackSessionInterfaceAVKit::modelDestroyed): Deleted.
(WebCore::PlaybackSessionInterfaceAVKit::logIdentifier const): Deleted.
(WebCore::PlaybackSessionInterfaceAVKit::loggerPtr const): Deleted.
(WebCore::PlaybackSessionInterfaceAVKit::logChannel const): Deleted.
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.h: Copied from Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.h.
* Source/WebCore/platform/ios/PlaybackSessionInterfaceIOS.mm: Added.
(WebCore::PlaybackSessionInterfaceIOS::PlaybackSessionInterfaceIOS):
(WebCore::PlaybackSessionInterfaceIOS::playbackSessionModel const):
(WebCore::PlaybackSessionInterfaceIOS::modelDestroyed):
(WebCore::PlaybackSessionInterfaceIOS::logIdentifier const):
(WebCore::PlaybackSessionInterfaceIOS::loggerPtr const):
(WebCore::PlaybackSessionInterfaceIOS::logChannel const):
* Source/WebCore/platform/ios/PlaybackSessionInterfaceLMK.h: Copied from Source/WebCore/platform/ios/PlaybackSessionInterfaceAVKit.h.
* Source/WebCore/platform/ios/PlaybackSessionInterfaceLMK.mm: Added.
(WebCore::PlaybackSessionInterfaceLMK::PlaybackSessionInterfaceLMK):
(WebCore::PlaybackSessionInterfaceLMK::~PlaybackSessionInterfaceLMK):
(WebCore::PlaybackSessionInterfaceLMK::durationChanged):
(WebCore::PlaybackSessionInterfaceLMK::currentTimeChanged):
(WebCore::PlaybackSessionInterfaceLMK::bufferedTimeChanged):
(WebCore::PlaybackSessionInterfaceLMK::rateChanged):
(WebCore::PlaybackSessionInterfaceLMK::seekableRangesChanged):
(WebCore::PlaybackSessionInterfaceLMK::canPlayFastReverseChanged):
(WebCore::PlaybackSessionInterfaceLMK::audioMediaSelectionOptionsChanged):
(WebCore::PlaybackSessionInterfaceLMK::legibleMediaSelectionOptionsChanged):
(WebCore::PlaybackSessionInterfaceLMK::externalPlaybackChanged):
(WebCore::PlaybackSessionInterfaceLMK::wirelessVideoPlaybackDisabledChanged):
(WebCore::PlaybackSessionInterfaceLMK::mutedChanged):
(WebCore::PlaybackSessionInterfaceLMK::volumeChanged):
(WebCore::PlaybackSessionInterfaceLMK::invalidate):
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.mm:
(VideoPresentationInterfaceIOS::create):
(VideoPresentationInterfaceIOS::VideoPresentationInterfaceIOS):
* Source/WebCore/platform/ios/WebAVPlayerController.h:
* Source/WebCore/platform/ios/WebAVPlayerController.mm:
(-[WebAVPlayerController playbackSessionInterface]):
(-[WebAVPlayerController setPlaybackSessionInterface:]):
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
(VideoFullscreenControllerContext::setUpFullscreen):
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionManagerProxy::createModelAndInterface):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(WKFullScreenViewControllerPlaybackSessionModelClient::setInterface):
* Source/WebKitLegacy/mac/WebView/WebVideoFullscreenController.mm:
(-[WebVideoFullscreenController init]):

Canonical link: https://commits.webkit.org/273994@main
